react/features/base/tracks/actions.js

```diff
@@ -52,6 +54,52 @@ export function destroyLocalTracks() {
 }
 
 /**
+ * Replaces one track with another for one renegotiation instead of invoking
+ * two renegotations with a separate removeTrack and addTrack. Disposes the
+ * removed track as well.
+ *
+ * @param {JitsiLocalTrack|null} oldTrack - The track to dispose.
+ * @param {JitsiLocalTrack|null} newTrack - The track to use instead.
+ * @param {JitsiConference} [conference] - The conference from which to remove
+ * and add the tracks. If one is not provied, the conference in the redux store
+ * will be used.
+ * @returns {Function}
+ */
+export function replaceLocalTrack(oldTrack, newTrack, conference) {
+    return (dispatch, getState) => {
+        const currentConference = conference
+            || getState()['features/base/conference'].conference;
+
+        return currentConference.replaceTrack(oldTrack, newTrack)
+            .then(() => {
+                // We call dispose after doing the replace because
+                //  dispose will try and do a new o/a after the
+                //  track removes itself.  Doing it after means
+                //  the JitsiLocalTrack::conference member is already
+                //  cleared, so it won't try and do the o/a
+                const disposePromise = oldTrack
+                    ? dispatch(_disposeAndRemoveTracks([ oldTrack ]))
+                    : Promise.resolve();
+
+                return disposePromise
+                    .then(() => {
+                        if (newTrack) {
+                            const muteAction = newTrack.isVideoTrack()
+                                ? setVideoMuted : setAudioMuted;
+
+                            return dispatch(muteAction(newTrack.isMuted()));
+                        }
+                    })
+                    .then(() => {
+                        if (newTrack) {
+                            return dispatch(_addTracks([ newTrack ]));
```

Comment:
I changed the order based on your feedback and what Saul said as well. Both of you agreed that adding the track then resetting mute made more sense. The reason I had mute before add is because the redux state for features/base/media also holds onto the current mute state and there is middleware that will cause any added track to match the mute state in features/base/media. So if you're muted and you add a track (such as by screensharing), the new track would remain muted. By setting mute before add, I was essentially setting up that middleware to no-op.

Making dispose -> add -> mute possible required promisify-ing some of the middleware and I'm unsure of the consequence. I'll test further but some help testing would be appreciated. Also, I haven't tested mobile yet with the promisify-ing commit.